Remove circular dependency warnings in ActionCable javascript and include source code in published package

rmacklin opened the pull request:
### Summary

In https://github.com/rails/rails/pull/34177, we've converted the ActionCable javascript to ES2015 (in that PR we avoided refactoring the existing structure, which has some circular dependency warnings). This PR builds on top of that work. In this PR we:

1. Reorganize the module imports to remove the circular dependency warnings
2. Update the ActionCable npm package to ship untranspiled ES2015 source code in addition to the heavier compiled bundle
   - This enables ActionCable users to ship smaller javascript bundles to their visitors with modern browsers. We added this enhancement to ActiveStorage in https://github.com/rails/rails/pull/31880.

In an [example repository](https://github.com/rmacklin/actioncable-es2015-build-example/tree/pr-34258), the bundle shrinks by 2.5K (23.6%) when you simply change the actioncable import to point to the untranspiled src, e.g. by [setting up an alias](https://github.com/rmacklin/actioncable-es2015-build-example/commit/172204aa043877158da131c92934b97ed5876dbb):

```
resolve: {
  alias: {
    actioncable: 'actioncable/src'
  }
}
```

If you go a step further, by [importing the tree-shaking-friendly source](https://github.com/rmacklin/actioncable-es2015-build-example/commit/f02de269be4c7ac98913a99689b126326e1f4e07), e.g.:
```diff
diff --git a/app/scripts/main.js b/app/scripts/main.js
index 17bc031..1a2b2e0 100644
--- a/app/scripts/main.js
+++ b/app/scripts/main.js
@@ -1,6 +1,6 @@
-import ActionCable from 'actioncable';
+import * as ActionCable from 'actioncable/action_cable';

 let cable = ActionCable.createConsumer('wss://cable.example.com');

 cable.subscriptions.create('AppearanceChannel', {
```

then the bundle shrinks by 3.2K (30%).

rmacklin committed:
Replace several ActionCable.* references with finer-grained imports

This reduces the number of circular dependencies among the module
imports from 4:

```
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/index.js
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection_monitor.js -> app/javascript/action_cable/index.js
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/consumer.js -> app/javascript/action_cable/index.js
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/subscriptions.js -> app/javascript/action_cable/index.js
```

to 2:

```
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/index.js
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/connection_monitor.js -> app/javascript/action_cable/index.js
```

rmacklin committed:
Extract ActionCable object to its own module

This is step in the direction of removing the circular dependencies.
We still have 2, but they have now changed from:

```
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/index.js
(!) Circular dependency: app/javascript/action_cable/index.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/connection_monitor.js -> app/javascript/action_cable/index.js
```

to:

```
(!) Circular dependency: app/javascript/action_cable/action_cable.js -> app/javascript/action_cable/consumer.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/action_cable.js
(!) Circular dependency: app/javascript/action_cable/action_cable.js -> app/javascript/action_cable/consumer.js -> app/javascript/action_cable/connection.js -> app/javascript/action_cable/connection_monitor.js -> app/javascript/action_cable/action_cable.js
```

While that might seem like all we did was introduce additional items
in the circular chain, we can now add additional named exports to this
newly added module, and that will allow us to get rid of the circular
dependencies.

rmacklin committed:
Reference INTERNAL module directly in ActionCable.createConsumer

This makes the module more self-contained because it doesn't have to
depend on someone else having set this.INTERNAL to the correct object.

rmacklin committed:
Extract getConfig and createWebSocketURL methods to exported functions

We're going to turn the ActionCable object into a series of
independently-importable exports so that we can remove the circular
dependencies.

We're starting with the lowest hanging fruit: `getConfig` and
`createWebSocketURL` do not reference any properties of the ActionCable
object, so they can simply be extracted out as regular functions that
we export.

rmacklin committed:
Extract createConsumer method of ActionCable to a function

After we extracted getConfig and createWebSocketURL, createConsumer no
longer referenced any properties of the ActionCable object, so it can
also be extracted to its own function that we export.

rmacklin committed:
Extract adapters module in ActionCable

This will help remove the remaining circular dependencies.

rmacklin committed:
Extract logger module in ActionCable

This removes the remainig circular dependencies, so there are now no
warnings when compiling actioncable

rmacklin committed:
Simplify ActionCable entry point

This moves the less important details to action_cable/action_cable.js so
that the entry point stays clean and simple.

rmacklin committed:
Include source code in published actioncable npm package

This allows actioncable users to ship smaller javascript bundles to
visitors using modern browsers, as demonstrated in this repository:
https://github.com/rmacklin/actioncable-es2015-build-example

In that example, the bundle shrinks by 2.8K (25.2%) when you simply
change the actioncable import to point to the untranspiled src.

If you go a step further, like this:
```
diff --git a/app/scripts/main.js b/app/scripts/main.js
index 17bc031..1a2b2e0 100644
--- a/app/scripts/main.js
+++ b/app/scripts/main.js
@@ -1,6 +1,6 @@
-import ActionCable from 'actioncable';
+import {createConsumer} from 'actioncable/action_cable';

-let cable = ActionCable.createConsumer('wss://cable.example.com');
+let cable = createConsumer('wss://cable.example.com');

 cable.subscriptions.create('AppearanceChannel', {
```

then the bundle shrinks by 3.6K (31.7%)!

In addition to allowing smaller bundles for those who ship untranspiled
code to modern browsers, including the source code in the published
package can be useful in other ways:

1. Users can import individual modules rather than the whole library
2. As a result of (1), users can also monkey patch parts of actioncable
   by importing the relevant module, modifying the exported object, and
   then importing the rest of actioncable (which would then use the
   patched object).

Note: This is the same enhancement that we made to activestorage in
c0368ad090b79c19300a4aa133bb188b2d9ab611

javan commented on actioncable/app/javascript/action_cable/action_cable.js:
```diff
@@ -0,0 +1,43 @@
+  logger.enabled = false
+}
+
+export { Connection, ConnectionMonitor, Consumer, INTERNAL, log, Subscription, Subscriptions }
```
Very nice! I like this approach to exporting the `ActionCable` interface much more than https://github.com/rails/rails/blob/ae122c32c11336734dc86d7bbdfdea5bfda2b6d6/actioncable/app/javascript/action_cable/index.js#L54-L61

The one thing that's missing is access to the `WebSocket` adapter here. e.g. being able to set `ActionCable.WebSocket = …` in an application.

rmacklin replied:
The syntax you’re referencing will still work if the application does:
```js
import ActionCable from "actioncable"
```
or uses the UMD export of the compiled bundle.

This behavior is verified by one of the existing tests: https://github.com/rails/rails/blob/405aef3dede8938c8eb9789f6d6722864faddb09/actioncable/test/javascript/src/unit/action_cable_test.coffee#L10-L12

If an application is using fine grained imports, they can opt to instead set the adapter via:
```js
import adapters from "actioncable/adapters"

adapters.WebSocket = MyAdapter
```

rmacklin replied:
Further conversation moved to https://github.com/rails/rails/pull/34177#discussion_r229398342

rmacklin replied:
Moving the discussion back here.

> > I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.
> 
> Can you share an example of that? How would you cherry-pick your imports for optimal tree-shaking with those separate files

I'm not sure if @sveinn has his own example, but the example repository I created demonstrates this: https://github.com/rmacklin/actioncable-es2015-build-example

Here's the way to cherry-pick the imports for optimal tree-shaking:
https://github.com/rmacklin/actioncable-es2015-build-example/commit/d494d6bf160042ca218bf3bd7aae6c1929492e0b#diff-4a23ac1286faa3273c8cdc9b4bb5078d

The difference in the unminified generated code is:
1. The assignment of `getConfig`, `createWebSocketURL`, `createConsumer`, `startDebugging`, `stopDebugging`, `Connection`, `ConnectionMonitor`, `Consumer`, `INTERNAL`, `log`, `Subscription`, `Subscriptions` onto `action_cable_namespaceObject` gets removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/831b2b90e6b0c072eabe1116b032b52d5735e416#diff-d2679f9a86538b3db8ad2a97579f226aL8
2. The delegated getters and setters for `ActionCable.logger` and `ActionCable.WebSocket` are not used in the application, so they get removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/831b2b90e6b0c072eabe1116b032b52d5735e416#diff-d2679f9a86538b3db8ad2a97579f226aL611

   And there are additional differences in when the optimized code is minified:

3. `createConsumer` is inlined into an anonymous function in the only place that it is used:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/57c4abe061ee6aad55e44d14592f59887f436566#diff-363285dc602b7237f78d880e26c875a0R348
4. `stopDebugging` and `startDebugging` are no longer referenced as a result of (1), so they get removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/57c4abe061ee6aad55e44d14592f59887f436566#diff-363285dc602b7237f78d880e26c875a0L29

rmacklin replied:
> and what would the difference be if they were combined?

If the files are combined you cannot preserve the backwards compatible API and also allow `createConsumer` to be imported individually. But I personally have no problem with changing the public API, for [these reasons](https://github.com/rails/rails/pull/34177#discussion_r229741614):

> However, if we don't want to have the two separate source files, there is an alternative that I would be fine with: We could also export `adapters` from `action_cable.js` and move `action_cable.js` to `index.js`. Then all functionality would be available from the individual exports of the new `index.js` (former `action_cable.js`), but we would be slightly changing the public API. I don't think that's a huge cause for concern because:
> 
> * We can share the contents of what is currently `index.js` in the upgrade guide as a way for users to patch the old API into their applications when they upgrade, until they migrate to the new API.
> * We could go further and ship a codemod that they can run to automatically translate their application to the new API. (If you're not familiar, this is something [react has been doing](https://github.com/reactjs/react-codemod) to make API upgrades easy for users.) I'd be happy to look into writing the codemod.

rmacklin replied:
@javan if I'm interpreting [your last comment](https://github.com/rails/rails/pull/34177#discussion_r229832626) correctly, it sounds like you're okay with modifying the public API by only offering individual exports instead of exporting an `ActionCable` object containing everything. Is that right?

If so I'll refactor this PR to remove `action_cable.js` and only offer individual exports from `index.js`.

rmacklin replied:
I've opened a PR with the alternate approach so we can easily compare them: https://github.com/rails/rails/pull/34370. I'm happy to close either PR in favor of the other, depending on which approach is preferred.

rmacklin committed:
Remove unused commonjs plugin from ActionCable rollup config

This added when we copied the rollup config from ActiveStorage, but
ActionCable does not have any commonjs dependencies (it doesn't have
any external dependencies at all), so this plugin is unnecessary here.

rmacklin commented:
The PR had a merge conflict in `actioncable/app/assets/javascripts/action_cable.js`. I fixed it but I rebased off of https://github.com/rails/rails/pull/34475 (not `master`) so that the generated bundle would be correct.

javan commented:
Hey, sorry for letting this sit so long! I'm still in favor of this change, but I think we shouldn't worry about preserving the existing API perfectly. It can be a breaking change (even though it doesn't break much) since the next release will be v6.

So, what do you think about doing all the individual `export`s in `index.js`, and defining ideal logger and WebSocket adapter APIs instead of backwards compatible ones.

javan commented:
Oops, forgot there were separate PRs. Yes, let's close this one and focus on https://github.com/rails/rails/pull/34370.
